### PR TITLE
Add cachedResolvingDecoder and SanityCheckBinaryDecoder and corresponding methods  

### PR DESCRIPTION
**Pull request by guawang**

### CODE REVIEW

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/SkipDecoder.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+package com.linkedin.avroutil1.compatibility;
+
+import java.io.IOException;
+import org.apache.avro.Schema;
+import org.apache.avro.io.Decoder;
+
+/**
+ * an abstract class that extends from Decoder that helps CachedResolvingDecoder
+ * perform various methods like original ResolvingDecoder.
+ */
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/SkipDecoder.java:16]: better documentation of what this is actually used for?

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/SkipDecoder.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+public abstract class SkipDecoder extends Decoder {
+
+  public int readStringSize() throws IOException {
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/SkipDecoder.java:19]: im assumming this reads the size of a string and is intended to help "skip" the actual string? if so  - it needs docs and also why have a default implementation at all?

File: helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java [hunk 3]
```diff
@@ -281,11 +281,23 @@ public static Decoder newCompatibleJsonDecoder(Schema schema, String in) throws
...
    * @throws IOException on io errors
    */
-  public static Decoder newCachedResolvingDecoder(Schema writer, Schema reader, Decoder in) throws IOException {
+  public static SkipDecoder newCachedResolvingDecoder(Schema writer, Schema reader, Decoder in) throws IOException {
     assertAvroAvailable();
     return ADAPTER.newCachedResolvingDecoder(writer, reader, in);
   }
 
+  /**
+   * {@link Decoder} that fixes a bug in the BinaryDecoder that can cause OutOfMemoryError
```
Comment (radai-rosenblatt) [inline helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java:290]: i think a better name is BoundedMemoryDecoder? i  assume all this does is protects against OOMs  when reading garbage?

File: helper/impls/helper-impl-110/src/main/java/com/linkedin/avroutil1/compatibility/avro110/codec/SanityCheckBinaryDecoder.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+import org.apache.avro.util.Utf8;
+
+
+/**
+ * Fixes a bug in the BinaryDecoder that can cause OutOfMemoryError when deserializing corrupt data or deserializing with the incorrect
+ * schema. Because this class uses the superclass constructor taking byte[], which is declared with package visiblilty,
+ * it has to be in the package org.apache.avro.io.
```
Comment (radai-rosenblatt) [inline helper/impls/helper-impl-110/src/main/java/com/linkedin/avroutil1/compatibility/avro110/codec/SanityCheckBinaryDecoder.java:17]: this doc is no longer true

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/SkipDecoder.java (hunk 2 above)
Comment (guawang) [reply helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/SkipDecoder.java:19]: will change all methods as abstract

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/SkipDecoder.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import org.apache.avro.Schema;
+import org.apache.avro.io.Decoder;
+
+/**
+ * An abstract class that extends from Decoder that supports skipping various
+ * symbol types and get the variable length of the fields.
+ */
+public abstract class SkipDecoder extends Decoder {
+
+  public abstract int readStringSize() throws IOException;
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/SkipDecoder.java:19]: needs doc - does this skip the  underlying input? does ti retrn the size of the fiels skipped? is that size in bytes? even for json? does it include the  size of payload alone, or also the size of the int  used to describe payload length?

Comment (radai-rosenblatt) [reply helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/SkipDecoder.java:19]: similarly for all methods here

**Review by radai-rosenblatt — APPROVED** [review]

**Review by radai-rosenblatt — APPROVED** [review]

**Pull request merged**